downstairs/src/extent_inner_raw_v2.rs

```diff
@@ -0,0 +1,1181 @@
+                Ok(Some(crucible_protocol::BlockContext {
+                    hash: b.hash,
+                    encryption_context: Some(
+                        crucible_protocol::EncryptionContext {
+                            tag: b.encryption_tag,
+                            nonce: b.encryption_nonce,
+                        },
+                    ),
+                }))
+            }
+            UNENCRYPTED_BLOCK_CONTEXT => {
+                Ok(Some(crucible_protocol::BlockContext {
+                    hash: b.hash,
+                    encryption_context: None,
+                }))
+            }
+            t => Err(CrucibleError::BadContextSlot(format!(
+                "invalid context tag: {t}"
+            ))),
+        }
+    }
+}
+
+/// `RawInnerV2` is a wrapper around a [`std::fs::File`] representing an extent
+///
+/// The file contains [`extent_size`](RawInnerV2::extent_size) blocks, which I'm
+/// writing as `N` for simplicity here.
+///
+/// # File organization
+/// The file is structured in three parts.  Getting specific offsets within the
+/// file is implemented in the [`RawLayout`] helper class.
+///
+/// ## Block data and contexts
+/// Block data and per-block contexts are paired up as follows:
+/// ```text
+/// [ ----- block  ----- | context | context | ----- block  ----- ] x (N / 2)
+/// ```
+///
+/// This packing lets us use half as many iovecs for large reads, because we can
+/// read contiguous pairs of contexts and blocks.
+///
+/// ## Written blocks array
+/// The next section of the file contains a bit-packed array indicating whether
+/// each block is written (where 0 is unwritten and 1 is written).  It takes up
+/// `N.div_ceil(8)` bytes.  It is only valid when the `dirty` bit is cleared.
+/// This is an optimization that speeds up opening a clean extent file;
+/// otherwise, we would have to read every block to find whether it has been
+/// written or not.
+///
+/// ## File metadata
+/// The last [`BLOCK_META_SIZE_BYTES`] in the file contain an [`OnDiskMeta`]
+/// serialized using `bincode`.  The first byte of this range is `dirty`,
+/// serialized as a `u8` (where `1` is dirty and `0` is clean).
+///
+/// There are a few considerations that led to this particular ordering:
+/// - The written blocks array and metadata must be contiguous, because we want
+///   to write them atomically when clearing the `dirty` flag.
+/// - We have multiple different raw file formats, but they all place an
+///   [`OnDiskMeta`] in the last [`BLOCK_META_SIZE_BYTES`] bytes of the file.
+///   This means we can read the metadata and pick the correct extent version.
+#[derive(Debug)]
+pub struct RawInnerV2 {
+    file: File,
+
+    /// Our extent number
+    extent_number: u32,
+
+    /// Extent size, in blocks
+    extent_size: Block,
+
+    /// Helper `struct` controlling layout within the file
+    layout: RawLayout,
+
+    /// Has this block been written?
+    block_written: Vec<bool>,
```

Comment:
Excuse my random outsider comment:

**suggestion:** This seems to never change in size after creation, so it could maybe be a boxed slice?